proposals/gc/MVP-JS.md

```diff
@@ -1,538 +1,65 @@
-  `StructType`/`ArrayType`, or should we have `WebAssembly.Function`
-  [as separately proposed](https://github.com/WebAssembly/js-types/blob/master/proposals/js-types/Overview.md#addition-of-webassemblyfunction)
-  (or both, with `WebAssembly.Function` subclassing `FunctionType`)?
-  Either way, the function type would be a [Type Definition object](#type-definition-objects),
-  whose signature was defined in terms of [Value Type objects](#value-type-objects).
-
-* How do [RTT values](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#runtime-types)
-  look when they flow out to JS? Are RTT values just first-class references to
-  [Typed Definition objects](#type-definition-objects)?
-
-* There is some hand-waving here around nullability / defaultability that needs
-  to be cleared up.
+The goal of the MVP JS API is to be as simple as possible, providing just enough
+expressivity to serve as the foundation for interop between JS and Wasm GC.
+Making it easy to access, manipulate, and extend Wasm GC objects from
+hand-written JS is left as future work. For now, require using the established
+pattern of exporting accessor functions to interact with Wasm data from JS.
+
+## Explanation
+
+WebAssembly GC objects are exposed to JS via opaque exotic object wrappers. The
+wrapper object internally holds a reference to the underlying WasmGC object to
+facilitate identity-preserving round-tripping from WebAssembly to JS and back.
+The wrapper object serves only to opaquely reference the underlying WebAssembly
+value and does not support inspecting or modifying the value in any way. The
+wrapper object also cannot be extended with new JS properties. The wrapper
+objects supports the following JS operations:
+
+ - `GetPrototypeOf`: returns `null`.
+ - `SetPrototypeOf`: returns `false` (and does not set a prototype).
+ - `IsExtensible`: returns `false`.
+ - `PreventExtensions`: returns `false`.
+ - `GetOwnProperty`: returns `undefined`.
+ - `DefineOwnProperty`: returns `false`.
+ - `HasProperty`: returns `false`.
+ - `Get`: returns `undefined`.
+ - `Set`: throws a `TypeError`.
+ - `Delete`: throws a `TypeError`.
+ - `OwnPropertyKeys`: returns an empty list.
+
+These operations may be extended in the future to support more useful ways of
+interacting with Wasm GC objects from JS.
+
+When a non-null internal reference value (i.e. a value with of any subtype of
+`anyref`) is passed out to JS, the `ToJSValue` algorithm will implicitly call
+`extern.externalize` on it to get a JS value. Similarly, when a JS value is
+passed into WebAssembly in a location that expects an internal reference, the
+`ToWebAssemblyValue` algorithm will implicitly call `extern.internalize` on it
+to get an internal reference. `ToWebAssemblyValue` will then cast the internal
+reference to the specific expected type, and if that fails, it will throw a
+`TypeError`.
+
+In JS hosts, `extern.internalize` will behave differently depending on the JS
```

Comment:
Side note: I believe it will be possible to specify this special-casing entirely at the boundary, and avoid a host-dependent patching of extern.{in/ex}ternalize. Conceptually, ToWasmValue can eagerly coerce small ints to the same representation as (externalised) Wasm i31 refs _before_ passing them on to internalize, so that Wasm already receives them as such and never needs to distinguish external values itself. But I can tweak that once that part of the core semantics is done.

Reply:
Makes sense. I'll add a TODO mentioning this as well.